### PR TITLE
NIFI-2032 fixed 'enableControllerServices' logic

### PR DESCRIPTION
**Pull request by olegz**
added getRequiredControllerServices() operation to ControllerServiceNode

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java [hunk 1]
```diff
@@ -70,7 +67,7 @@
 
     private final ProcessScheduler processScheduler;
     private static final Set<Method> validDisabledMethods;
-    private final BulletinRepository bulletinRepo;
+    // private final BulletinRepository bulletinRepo;
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: Is the `BulletinRepository` no longer necessary?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java [hunk 2]
```diff
@@ -372,98 +369,32 @@ public void enableControllerService(final ControllerServiceNode serviceNode) {
...
-                t.setName("Enable Controller Services");
-                return t;
-            }
-        });
-
-        for (final List<ControllerServiceNode> branch : branches) {
-            final Runnable enableBranchRunnable = new Runnable() {
+        if (shouldStart) {
+            List<ControllerServiceNode> services = new ArrayList<>(serviceNodes);
+            Collections.sort(services, new Comparator<ControllerServiceNode>() {
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: This looks like a much cleaner way to perform this behavior, but are there any tests for the dependency traversal to ensure it's complete?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java (hunk 1 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: I assume so, this change is du to changes from Mark and Matt this morning. I probably should remove comment

Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: I removed the two commented lines (the declaration and the assignment in the constructor). However, I did not change the constructor contract, as it is used 17 times throughout the application. @mcgilman @markap14 is the bulletin repository work going to have additional impact, or is it safe to clean up this constructor universally?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java [hunk 3]
```diff
@@ -372,98 +369,32 @@ public void enableControllerService(final ControllerServiceNode serviceNode) {
...
-                                try {
-                                    serviceNode.verifyCanEnable();
-                                    processScheduler.enableControllerService(serviceNode);
-                                } catch (final Exception e) {
-                                    logger.error("Failed to enable " + serviceNode + " due to " + e);
-                                    if (logger.isDebugEnabled()) {
-                                        logger.error("", e);
-                                    }
-
-                                    if (bulletinRepo != null) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: Looks like we were creating a bulletin whenever we were unable to enable a Controller Service.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java (hunk 1 above)
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: I would have to defer to @markap14 but it appears it was being used. See below.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java (hunk 3 above)
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: Ok, we should probably maintain that behavior, correct? Not sure why it was commented out initially.

**Comment by alopresto** [comment]
Thanks @olegz . I will build locally and merge.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java (hunk 3 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java]: So, I've just put _bulletinRepo_ logic back. The commit isn't squashed so you can see the changes in isolation.

**Comment by alopresto** [comment]
Getting a compilation failure missing `import org.apache.nifi.reporting.BulletinFactory`. Did this get missed in the second commit?

**Comment by olegz** [comment]
I see, it appears that I have to reference different ```org.apache.nifi.events.BulletinFactory```, building now

**Comment by olegz** [comment]
@alopresto so I just fixed it and squashed it with the last commit. No compilation error on my end. LMK

**Comment by alopresto** [comment]
I merged into master but I've had to rebase and resolve conflicts 3 times on `0.x` branch and still have tests failing. Will continue working on that part in the morning.

**Comment by olegz** [comment]
@alopresto let me look at what's going on with 0.x and I'll issue a separate PR. Thanks!

**Pull request closed without merging**